### PR TITLE
feat(middleware) Pass params to process_resource

### LINKED ISSUE #612: Pass params to process_resource
This is needed for feature parity with global *before* hooks. We can use a shim to avoid breaking existing middleware.

### PR DESCRIPTION
**Pull request by yohanboniface**
This allows middlewares to interact with route parameters.

BREAKING CHANGE: `process_resource` must now accept `params`
argument. Fix #612

I've arbitrated in favour of an API breaking change, because I think it's better to break API now (before 1.0) than to support useless legacy tricks, but of course this is to be discussed :)

### CODE REVIEW

File: tests/test_middlewares.py [hunk 1]
```diff
@@ -55,9 +55,17 @@ def process_request(self, req, resp):
         req.path = req.path.replace('/base_path', '', 1)
 
 
+class AccessParamsMiddleware(object):
+
+    def process_resource(self, req, resp, resource, params):
```
Comment (kgriffs) [inline tests/test_middlewares.py]: I think we should also test to ensure that a parameter that is part of the path shows up here in params.

File: tests/test_middlewares.py [hunk 2]
```diff
@@ -55,9 +55,17 @@ def process_request(self, req, resp):
...
+        global context
+        params['added'] = True
```
Comment (kgriffs) [inline tests/test_middlewares.py]: Would you mind adding a test to check that the modified parameter is getting propagated correctly to the responder method on the resource?

**Comment by kgriffs** [comment]
I think this makes sense. It is a breaking change, but one that should be easy for developers to accommodate. A couple testing suggestions inline.

File: tests/test_middlewares.py (hunk 1 above)
Comment (yohanboniface) [inline tests/test_middlewares.py]: Done :)

File: tests/test_middlewares.py (hunk 2 above)
Comment (yohanboniface) [inline tests/test_middlewares.py]: Done :)

File: falcon/api.py [hunk 3]
```diff
@@ -544,13 +545,13 @@ def _call_req_mw(self, stack, req, resp):
...
 
-    def _call_rsrc_mw(self, stack, req, resp, resource):
+    def _call_rsrc_mw(self, stack, req, resp, resource, params):
         """Run process_resource middleware methods."""
 
         for component in self._middleware:
             _, process_resource, _ = component
             if process_resource is not None:
-                process_resource(req, resp, resource)
+                process_resource(req, resp, resource, params)
```
Comment (philiptzou) [inline falcon/api.py]: I am not quite sure if detecting `process_resource` with `inspect.getargspec` is a good idea (a bit downside about the performance, huh?). But with that technique we can easily prevent the breaking change.

Comment (kgriffs) [inline falcon/api.py]: One way to mitigate the performance issue is to do the work up front in `prepare_middleware` to detect the number of args, and put a shim in place as needed for older middleware components. E.g. (pseudocode):

```
if is old-style method:
    @functools.wraps(orig_process_resource)
    def func(req, resp, resource, params)):
         orig_process_resource(req, resp, resource)

    process_resource = func
else:
    process_resource = orig_process_resource
```

Then the overhead is reduced to an extra function call. Eventually we can remove the shim as a breaking change once everyone has had a chance to migrate (e.g., 2.0).

What do you think?

File: tests/test_middlewares.py [hunk 4]
```diff
@@ -368,3 +376,23 @@ def test_base_path_is_removed_before_routing(self):
...
+
+class TestResourceMiddleware(TestMiddleware):
+
+    def test_can_access_resource_params(self):
+        """Test that params can be accessed from within process_resource"""
+        global context
+
+        class Resource:
+            def on_get(self, req, resp, **params):
+                resp.body = params
```
Comment (kgriffs) [inline tests/test_middlewares.py]: Nitpick: The fact that this works without serializing to a string is an implementation detail that may or may not hold up in the future. Instead, you may want to pass `params` through `json.dumps()`, assigning the result to `resp.data`.

File: falcon/api.py (hunk 3 above)
Comment (yohanboniface) [inline falcon/api.py]: Are we sure we want to take this path?
In my opinion, 0.x versions are made to keep maintainers out from 100% code backward compatibility. And specifically is this case we are only talking about adding a parameter to "process_resource" middlewares methods. It does not affect any logic or API behaviour. So instead, I would just add this breaking change in the release note, and let falcon users *just* go through every middleware and add an argument.
In other words, I think the balance here, in this specific case, is in favour of adding a bit of charge to falcon users instead of adding more code complexity and legacy on the shoulders of the maintainer.
Thoughts?

File: tests/test_middlewares.py (hunk 4 above)
Comment (yohanboniface) [inline tests/test_middlewares.py]: Done :)

File: falcon/api.py (hunk 3 above)
Comment (kgriffs) [inline falcon/api.py]: All things considered, I think it's OK to make the breaking change. There will be very few breaking changes in the next release as it is, and this one is easy enough to work around. Let's go ahead and land this for 0.4. That will be the final release before 1.0.

**Comment by kgriffs** [comment]
Looks great, thanks!
